streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java

```diff
@@ -0,0 +1,348 @@
+        }
+        boolean cyclePossible = true;
+        while (cyclePossible) {
+            cyclePossible = false;
+            for (final V node : nodes) {
+                final Map<V, V> parentNodes = new HashMap<>();
+                final Map<V, Edge> parentEdges = new HashMap<>();
+                final V nodeInCycle = detectNegativeCycles(node, parentNodes, parentEdges);
+                if (nodeInCycle == null) {
+                    continue;
+                }
+                cyclePossible = true;
+                V parentNode = parentNodes.get(nodeInCycle);
+                Edge parentEdge = parentEdges.get(nodeInCycle);
+
+                // Find max possible negative flow
+                int possibleFlow = parentEdge.residualFlow;
+                for (V curNode = parentNode; curNode != nodeInCycle; curNode = parentNodes.get(curNode)) {
+                    parentEdge = parentEdges.get(curNode);
+                    possibleFlow = Math.min(possibleFlow, parentEdge.residualFlow);
+                }
+
+                // Update graph by removing negative flow
+                parentNode = parentNodes.get(nodeInCycle);
+                parentEdge = parentEdges.get(nodeInCycle);
+                Edge counterEdge = parentEdge.counterEdge;
+                parentEdge.residualFlow -= possibleFlow;
+                parentEdge.flow += possibleFlow;
+                counterEdge.residualFlow += possibleFlow;
+                if (counterEdge.flow >= possibleFlow) {
+                    counterEdge.flow -= possibleFlow;
+                }
+                for (V curNode = parentNode; curNode != nodeInCycle; curNode = parentNodes.get(curNode)) {
+                    parentEdge = parentEdges.get(curNode);
+                    counterEdge = parentEdge.counterEdge;
+                    parentEdge.residualFlow -= possibleFlow;
+                    parentEdge.flow += possibleFlow;
+                    counterEdge.residualFlow += possibleFlow;
+                    if (counterEdge.flow >= possibleFlow) {
+                        counterEdge.flow -= possibleFlow;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Detect negative cycle using Bellman-ford shortest path algorithm
+     * @param source Source node
+     * @param parentNodes Parent nodes to store negative cycle nodes
+     * @param parentEdges Parent edges to store negative cycle edges
+     *
+     * @return One node in negative cycle if exists or null if there's no negative cycle
+     */
+    private V detectNegativeCycles(final V source, final Map<V, V> parentNodes, final Map<V, Edge> parentEdges) {
+        // Use long to account for any overflow
+        final Map<V, Long> distance = nodes.stream().collect(Collectors.toMap(node -> node, node -> (long) Integer.MAX_VALUE));
+        distance.put(source, 0L);
+        final int nodeCount = nodes.size();
+
+        for (int i = 0; i < nodeCount; i++) {
```

Comment:
From https://en.wikipedia.org/wiki/Bellman%E2%80%93Ford_algorithm it says, so it N-1 times, but we do it N times. Why?

Reply:
Good question. N-1 times will find shortest path, we need to do it one more time to see if the path can be even shorter which mean there's a negative cycle. That's why there's a check on line 335.